byterun/minor_gc.c

```diff
@@ -170,6 +176,56 @@ void caml_set_minor_heap_size (asize_t bsz)
   reset_table ((struct generic_table *) &caml_custom_table);
 }
 
+mlsize_t caml_distance_to_closure_environment(value closure)
+{
+#ifndef NATIVE_CODE_AND_NO_NAKED_POINTERS
+  return 0;
+#else
+  int finished = 0;
+  value* start;
+  value* pos;
+  mlsize_t skip;
+
+  CAMLassert(Is_block(closure));
+  CAMLassert(Tag_val(closure) == Closure_tag);
+  CAMLassert(Wosize_val(closure) >= 2);
+
+  start = (value*) closure;
+  pos = start;
+
+  while (!finished) {
+    uintnat arity;
+
+    CAMLassert ((pos - start) < Wosize_val(start));
+
+    arity = Long_val(pos[1]);
+    finished = (arity & (uintnat) 1) != 0;
```

Comment:
This should be 'documented' as an arity manipulating macro (or static inline function), probably in `mlvalues.h`.
`#define Last_function_of_closures(x) ...`